### ISSUE TITLE
Frustum association

### ISSUE THREAD

**Issue by CGL-learner** [mrnabati/CenterFusion#91]
Does anyone could provide the Frustum association part code? I don't find the code of this part.

**Comment by trainmker** [comment]
**Path of file: CenterFusion/src/lib/utils/pointcloud.py**
**Bold part** in below code is where the frustum association occurs. it is an **approximate of 3d frustum not exact 3d mapping** from what i got to understand. 
def generate_pc_hm(output, pc_dep, calib, opt):
      K = opt.K
      # K = 100
      heat = output['hm']
      wh = output['wh']
      pc_hm = torch.zeros_like(pc_dep)

      batch, cat, height, width = heat.size()
      scores, inds, clses, ys0, xs0 = _topk(heat, K=K)
      xs = xs0.view(batch, K, 1) + 0.5
      ys = ys0.view(batch, K, 1) + 0.5
      
      ## Initialize pc_feats
      pc_feats = torch.zeros((batch, len(opt.pc_feat_lvl), height, width), device=heat.device)
      dep_ind = opt.pc_feat_channels['pc_dep']
      vx_ind = opt.pc_feat_channels['pc_vx']
      vz_ind = opt.pc_feat_channels['pc_vz']
      to_log = opt.sigmoid_dep_sec
      
      ## get estimated depths
      out_dep = 1. / (output['dep'].sigmoid() + 1e-6) - 1.
      dep = _tranpose_and_gather_feat(out_dep, inds) # B x K x (C)
      if dep.size(2) == cat:
        cats = clses.view(batch, K, 1, 1)
        dep = dep.view(batch, K, -1, 1) # B x K x C x 1
        dep = dep.gather(2, cats.long()).squeeze(2) # B x K x 1

      ## get top bounding boxes
      wh = _tranpose_and_gather_feat(wh, inds) # B x K x 2
      wh = wh.view(batch, K, 2)
      wh[wh < 0] = 0
      if wh.size(2) == 2 * cat: # cat spec
        wh = wh.view(batch, K, -1, 2)
        cats = clses.view(batch, K, 1, 1).expand(batch, K, 1, 2)
        wh = wh.gather(2, cats.long()).squeeze(2) # B x K x 2
      bboxes = torch.cat([xs - wh[..., 0:1] / 2, 
                          ys - wh[..., 1:2] / 2,
                          xs + wh[..., 0:1] / 2, 
                          ys + wh[..., 1:2] / 2], dim=2)  # B x K x 4
      
      ## get dimensions
      dims = _tranpose_and_gather_feat(output['dim'], inds).view(batch, K, -1)

      ## get rotation
      rot = _tranpose_and_gather_feat(output['rot'], inds).view(batch, K, -1)

      ## Calculate values for the new pc_hm
      clses = clses.cpu().numpy()

      for i, [pc_dep_b, bboxes_b, depth_b, dim_b, rot_b] in enumerate(zip(pc_dep, bboxes, dep, dims, rot)):
        alpha_b = get_alpha(rot_b).unsqueeze(1)

        if opt.sort_det_by_dist:
          idx = torch.argsort(depth_b[:,0])
          bboxes_b = bboxes_b[idx,:]
          depth_b = depth_b[idx,:]
          dim_b = dim_b[idx,:]
          rot_b = rot_b[idx,:]
          alpha_b = alpha_b[idx,:]

 ****for j, [bbox, depth, dim, alpha] in enumerate(zip(bboxes_b, depth_b, dim_b, alpha_b)):
clss = clses[i,j].tolist()
ct = torch.tensor([(bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2], device=pc_dep_b.device)
dist_thresh = get_dist_thresh(calib, ct, dim, alpha)
dist_thresh += dist_thresh * opt.frustumExpansionRatio
pc_dep_to_hm_torch(pc_hm[i], pc_dep_b, depth, bbox, dist_thresh, opt)****
      return pc_hm
